xDS Server Implementation

banks opened the pull request:
**Note: this PR targets merging into f-envoy and depends directly on #4730**. This is part of a series of PRs that were developed together but split for easier review.

This is the actual xDS API server implementation. It authenticates gRPC streams, fetches config and watches for changes from the `proxycfg.Manager` from #4729, and then delivers updates to Envoy for  TLS, upstream listeners and service discovery results.

## Overview

![proxy config manager](https://user-images.githubusercontent.com/120915/46301077-397f9500-c59d-11e8-8b2c-9803452eb748.png)

## Notably Missing

 - xDS server is not wired up to the agent in this PR, that, the `connect envoy` CLI command and some other plumbing will be in a separate PR next. End to end tests aren't really possible at this point so I'll wait.
 - Documentation: this is left for later since the whole of f-envoy branch needs a large docs overhaul. That will all be addressed separately once this code is in review.

mitchellh commented on agent/xds/server.go:
```diff
@@ -0,0 +1,418 @@
+
+// NewServer creates a usable server instance.
+func NewServer(logger *log.Logger, cfgMgr ConfigManager,
+	aclFn ACLResolverFunc, authz ConnectAuthz) *Server {
```
😅 Config struct, I think.

banks replied:
Ha ha in this case the Server struct is _only_ config and has nothing else to initialize so the NewServer func can just go away I think.

mitchellh replied:
YAY

mitchellh commented on agent/xds/server.go:
```diff
@@ -0,0 +1,418 @@
+// ConfigManager is the interface the config manager implements to make testing
+// the server without a full mocker manager, state and cache. In normal use this
+// is satisfied directly by a pointer to the agent's proxycfg.Manager.
+type ConfigManager interface {
```
Can you also explain what a ConfigManager is? To the new reviewer, it isn't obvious what it does until reading more.

banks replied:
Hmm I tried to, it's just a shim to allow us to test without needing a full stack of mocked cache, local state and proxy manager. I'll try to reword it if that wasn't clear. Maybe it should read:

> ConfigManager is the interface we require to consume proxy config updates. It's satisfied normally by the agent's proxycfg.Manager, but allows easier testing without several layers of mocked cache, local state and proxycfg.Manager.

mitchellh commented on agent/xds/server.go:
```diff
@@ -0,0 +1,418 @@
+			// a normal shutdown where client goes away too.
+			return nil
+		case req, ok = <-reqCh:
+			if !ok {
```
Do we even really need to close the channel out and handle this? Or can we just block forever because `stream.Context().Done()` is always set? 

I don't really care, but just wondering because usually its not that critical to close channels unless the receiver needs to really know its closed. But in this case we seem to have an authoritative way to cancel.

banks replied:
I'm not sure is the honest answer and was playing it safe.

The check here is because we close this channel if the `Recv` fails on the stream in another goroutine - I'm not sure if we are guaranteed that gRPC will always cancel the stream's context in that case but we should abort processing the RPC as we can't make progress any more.

gRPC guarantees to cancel the stream context once the RPC has returned which we can't observe here since this method blocks the RPC method), I included the case here in case it also does it earlier if the client hangs up or similar. I expect we'd see that via a Recv fail too/instead, just couldn't find clarity on what guarantees are made about the ordering there so just handled everything that might happen!

Even the code in the official XDS server SDK has TODO comments to figure out how gRPC cancels stuff.

So yeah I'd like to spend more time reading gRPC code to get this right, but this is a defensive approach in liu of having time to investigate more 😄.

banks replied:
According to https://github.com/grpc/grpc-go/issues/465 it looks like the stream context is only ever cancelled on the server side once the RPC handler returns so the this is actually necessary to be signalled to stop on a Recv error (which is how we know if the client drops).

I'll add a comment and remove the context case here since it seems like it can never be hit now.

rboyer commented on agent/cache/watch.go:
```diff
@@ -0,0 +1,122 @@
+		return fmt.Errorf("watch requires the type to support blocking")
+	}
+
+	// Always start at 0 index to deliver the inital (possibly currently cached
```
typo `s/inital/initial/`

rboyer commented on agent/proxycfg/manager.go:
```diff
@@ -0,0 +1,273 @@
+// certificates, intentions, and service discovery results for any declared
+// upstreams.
+type Manager struct {
+	// cache is the agent's cache instance that can be used to retreive, store and
```
typo: `s/retreive/retrieve/`

rboyer commented on agent/xds/clusters.go:
```diff
@@ -0,0 +1,79 @@
+
+	"github.com/gogo/protobuf/proto"
+
+	"github.com/envoyproxy/go-control-plane/envoy/api/v2"
```
`v2`, `auth`, `core` would probably read a little better in the body of the file if they packages imports were aliased to something like `v2->envoy`, `auth->envoyauth`, `core->envoycore`

rboyer commented on agent/xds/listeners.go:
```diff
@@ -0,0 +1,229 @@
+	var l v2.Listener
+	err = jsonpb.UnmarshalString(configJSON, &l)
+	if err != nil {
+		return nil, err
```
is there value in retaining the first error from the `jsonpb` unmarshal failure and returning that here too in a multierror?

rboyer commented on agent/xds/listeners.go:
```diff
@@ -0,0 +1,229 @@
+	// as JSON and decoding again!!
+	cfgStruct, err := util.MessageToStruct(cfg)
+	if err != nil {
+		panic(err)
```
So practically speaking I'm guessing this is never going to happen?

rboyer commented on agent/xds/server.go:
```diff
@@ -0,0 +1,418 @@
+	reqStop := int32(0)
+	go func() {
+		for {
+			req, err := stream.Recv()
```
Does this block? If so how do you coax it to stop during server shutdown?

banks replied:
Yes, but the gRPC server controls it and is the thing that unblocks it with an error (EOF?) when the client is disconnected. Presumably that includes when the server terminates. This appears to be the expected pattern for gRPC streaming...

rboyer commented on agent/xds/server.go:
```diff
@@ -0,0 +1,418 @@
+		case stateInit:
+			if req == nil {
+				// This can't happen (tm) since stateCh is nil until after the first req
+				// is recieved but lets not panic about it.
```
typo: `s/recieved/received/`

rboyer commented on agent/xds/server.go:
```diff
@@ -0,0 +1,418 @@
+
+			// Start watching config for that proxy
+			stateCh, watchCancel = s.cfgMgr.Watch(proxyID)
+			defer watchCancel()
```
Did you mean to do a `defer` inside of the wrapped `for` loop? Usually that can backfire.  It's hard to tell from the flow of this state machine if this is ok in context.

banks replied:
Yep, comment added.

rboyer commented on agent/xds/server.go:
```diff
@@ -0,0 +1,418 @@
+			// range the map which has no determined order. It's important because:
+			//
+			//  1. Envoy needs to see a consistent snapshot to avoid potentially
+			//     dropping or misrouting traffic due to inconsitencies. This is the
```
typo: `s/inconsitencies/inconsistencies/`

rboyer commented on agent/xds/server.go:
```diff
@@ -0,0 +1,418 @@
+			//     to handle
+			for _, typeURL := range []string{ClusterType, EndpointType, RouteType, ListenerType} {
+				handler := handlers[typeURL]
+				err := handler.SendIfNew(cfgSnap, configVersion, &nonce)
```
trivial: you can collapse these two lines into into: `if err := ...; err != nil {`

banks replied:
I did this although it puts the line 1 char over our typical style guideline... 😄

rboyer commented on agent/xds/server.go:
```diff
@@ -0,0 +1,418 @@
+	// ordered as far as I can tell, but it's cheaper just to increment a counter
+	// every time we observe a new config since the upstream proxycfg package only
+	// delivers updates when there are actual changes.
+	var configVersion uint64
```
How does this whole version dance work if Consul is restarted while Envoy remains running?

banks replied:
Great question.

I think this is safe right now due to the way `xDSType.lastVersion` is handled lower down this file. I'll walk through an example to convince myself. (This is more by coincidence than good design.)

 - Assume Envoy and agent are up and Envoy has loaded version 3 already.
 - Agent restarts
 - Envoy reconnects with same discovery request showing it's current "version = 3"
 - Agent sees the current config snapshot after Watch was called which it assigns `version 1` due to having restarted
 - It delivers this
 - SendIfNew compares with `t.lastVersion` which is 0 since that was also just re-initialised and this is only the last version we _sent_ not updated by the incoming request
   - Since 1 > 0, the version is sent
 - Envoy applies the config and is now on `version 1` and resumes normally. 
 - We set t.lastVersion to be 1 and continue normally.

I think there is one corner cases where it might be problematic though:

If envoy is on version 1, the agent is stopped, the proxy config altered and then started again.
    - In this case the new config will be sent to Envoy but since it is also "version 1" Envoy _might_ ignore it as the same as the config it has applied (need to verify).

I think it's OK for now on the basis that I took this exact pattern from the xDS SDK and reference implementation. I assume based on that it's either not a real problem (because Envoy will re-apply config even if version is same as current) or is a huge edge case that's not a big deal yet. We can revisit later I guess.

banks replied:
I tested it seems like Envoy still applies a config even if it's version is the same as the current one. In my case I started agent, let it deliver initial certs (at version 1) then restarted agent. When envoy reconnects it gets and loads a new set of certs even though it's still v1. So we are good for now.

rboyer replied:
If you wanted more belt-and-suspenders you could either persist the last version to disk on the agent and pick up where you left off, or see if you could have the agent generate a random token on startup and see if there's a way to give that to the envoy process to strongly indicate "I'm controlled by a new master, so I should assume it's all different now". Sorta like a raft term.

rboyer replied:
Or perhaps have the agent interrogate envoy to see what the loaded version is, and then internally sync the agent so that the next version it hands out is `envoyLoadedVersion+1`

rboyer commented on agent/xds/server.go:
```diff
@@ -0,0 +1,418 @@
+
+// IncrementalAggregatedResources implements discovery.AggregatedDiscoveryServiceServer
+func (s *Server) IncrementalAggregatedResources(_ discovery.AggregatedDiscoveryService_IncrementalAggregatedResourcesServer) error {
+	return errors.New("not implemented")
```
What are the implications of implementing-but-not-implementing this method?

banks replied:
It's an optional API that Envoy has to be configured to use currently. We should totally implement it in the future as it's more scalable that reloading the entire state every time one thing changes but not necessary as it's an optional protocol extension.

rboyer commented on agent/xds/server.go:
```diff
@@ -0,0 +1,418 @@
+// requests.
+func (s *Server) GRPCServer(certFile, keyFile string) (*grpc.Server, error) {
+	opts := []grpc.ServerOption{
+		grpc.MaxConcurrentStreams(2048),
```
Why 2048?

banks replied:
Seems like enough :)

The docs are super unclear but it was received wisdom for a while at least that gRPC limits to 100 streams by default which is probably enough for us but just in case you have more than 100 envoys talking to one agent I bumped it up. For example: https://github.com/cockroachdb/cockroach/issues/14769 

The current code doesn't _seem_ to default to 100 anywhere from a quick look but this was just a reasonable number in general.

rboyer commented on agent/proxycfg/manager.go:
```diff
@@ -0,0 +1,273 @@
+				continue
+			}
+			// TODO(banks): need to work out when to default some stuff. For example
+			// Proxy.LocalServicePort is practially necessary for any sidecar and can
```
typo: `s/practially/practically/`

rboyer commented on agent/proxycfg/manager.go:
```diff
@@ -0,0 +1,273 @@
+	}
+}
+
+// ensureProxyServiceLocked adss or changes the proxy to our state.
```
typo: `s/adss/adds/`

rboyer commented on agent/proxycfg/manager.go:
```diff
@@ -0,0 +1,273 @@
+			return nil
+		}
+
+		// We are updating the proxy, close it's old state
```
`s/it's/its/`

rboyer replied:
`s/it's/its/`

rboyer commented on agent/proxycfg/manager.go:
```diff
@@ -0,0 +1,273 @@
+		}
+
+		// We are updating the proxy, close it's old state
+		state.Close()
```
What about the `error` this returns?

rboyer commented on agent/proxycfg/manager.go:
```diff
@@ -0,0 +1,273 @@
+		ch := state.Watch()
+		for {
+			select {
+			case snap, ok := <-ch:
```
If you only have the one case, why bother with the `select` at all? You could just do a bare `snap, ok := <-ch`

rboyer commented on agent/proxycfg/manager.go:
```diff
@@ -0,0 +1,273 @@
+
+	// Closing state will let the goroutine we started in Ensure finish since
+	// watch chan is closed.
+	state.Close()
```
handle the error?

rboyer commented on agent/proxycfg/manager.go:
```diff
@@ -0,0 +1,273 @@
+		// TODO(banks): should we close their chan here to force them to eventually
+		// notice they are too slow? Not sure if it really helps.
+		select {
+		case ch <- snap:
```
Is this the thing I already commented on in the prior PR with the head-swapping of the topmost item in a chan-of-size-1?

rboyer commented on agent/proxycfg/manager.go:
```diff
@@ -0,0 +1,273 @@
+	return nil
+}
+
+// removeProxyService is called when a service deregisteres and frees all
```
`s/deregisteres/deregisters/`

banks commented:
Last commits addressed most of the points here which were awesome.

I also realised that the "let the user give us custom config" was a bit too liberal to be useful at all - they would also not get Connect certs or Authz that way!

The new behaviour implemented (and tested) in last few commits is to override the TLS config and insert the ext_authz filter onto the start of every configured filter chain even ones coming from the user. This at least ensures their main public listener as advertised in Consul will have the right TLS configs and with do Auth correctly before anything else.

I also realised we probably need to allow customisation of the Cluster config too for some features but we can iterate on that after this PR/initial release.
